x/poolmanager/client/query_proto_wrap.go

```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
+	)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+		InputCoin:  currFromCoin,
+		OutputCoin: tokenOut,
+	}, nil
+}
+
+// estimateTradeBasedOnPriceImpactConcentratedLiquidity estimates a trade based on price impact for a concentrated
+// liquidity pool type. For a concentrated liquidity pool if an amount entered is greater than the total pool liquidity
+// the trade estimated would error. If the amount is small it would return tokenOut to be 0 in which case we should
+// return an empty trade. If the estimate returns an error we should ignore it and continue attempting to estimate
+// by halving the input.
+func (q Querier) estimateTradeBasedOnPriceImpactConcentratedLiquidity(
+	ctx sdk.Context,
+	req queryproto.EstimateTradeBasedOnPriceImpactRequest,
+	spotPrice, adjustedMaxPriceImpact sdk.Dec,
+	swapModule types.PoolModuleI,
+	poolI types.PoolI,
+) (*queryproto.EstimateTradeBasedOnPriceImpactResponse, error) {
+	tokenOut, err := swapModule.CalcOutAmtGivenIn(ctx, poolI, req.FromCoin, req.ToCoinDenom, sdk.ZeroDec())
+	// If there was no error we attempt to validate if the output is below the adjustedMaxPriceImpact.
+	if err == nil {
+
+		// If the tokenOut was returned to be zero it means the amount being traded is too small. We ignore the
+		// error output here as it could mean that the input is too large.
+		if tokenOut.IsZero() {
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+				OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+			}, nil
+		}
+
+		currTradePrice := sdk.NewDec(req.FromCoin.Amount.Int64()).QuoInt(tokenOut.Amount)
+		priceDeviation := currTradePrice.Sub(spotPrice).Quo(spotPrice).Abs()
+
+		if priceDeviation.LTE(adjustedMaxPriceImpact) {
+			tokenOut, err = swapModule.CalcOutAmtGivenIn(
+				ctx, poolI, req.FromCoin, req.ToCoinDenom, poolI.GetSpreadFactor(ctx),
+			)
+			if err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  req.FromCoin,
+				OutputCoin: tokenOut,
+			}, nil
+		}
+	}
+
+	// Define low and high amount to search between. Start from 1 and req.FromCoin.Amount as initial range.
+	lowAmount := sdk.OneInt()
+	highAmount := req.FromCoin.Amount
+	currFromCoin := req.FromCoin
+
+	for lowAmount.LTE(highAmount) {
+		// Calculate currFromCoin as the new middle amount to try trade.
+		midAmount := lowAmount.Add(highAmount).Quo(sdk.NewInt(2))
+		currFromCoin = sdk.NewCoin(req.FromCoin.Denom, midAmount)
+
+		tokenOut, err := swapModule.CalcOutAmtGivenIn(ctx, poolI, currFromCoin, req.ToCoinDenom, sdk.ZeroDec())
+		if err == nil {
+			// If the tokenOut was returned to be zero it means the amount being traded is too small. We ignore the
+			// error output here as it could mean that the input is too large.
```

Comment:
ditto here for method generalization and removing code duplication

Reply:
I've started doing it in one algorithm but it becomes very hard to follow because each pool reacts differently with errors and panics. I can change it but I suggest against it.